### DISCUSSION TITLE
Use SD Card with the AVR IoT Cellular Mini

### DISCUSSION THREAD

**Discussion by vonrle** [microchip-pic-avr-solutions/avr-iot-cellular-arduino-library#19] [Q&A]
Hello everyone, 

I tried to connect a Feather SD Card Device with my AVR IoT Cellular Mini Board and have some issues with it. From the documentation I know, that the Feather CS Pin is the Pxn, but I couldn't find the correct way to call that PIN, as I get an error, when I use "PIN_Pxn" in my code, that PIN_Pxn needs to be defined. 

I would be very greatful for your help!

**Comment by M65649** [comment]
Hello @vonrle,

Could you show some code?

**Comment by vonrle** [comment]
Sure, I just used the example code of the SD Card library: 
`/*
  SD card datalogger

  This example shows how to log data from three analog sensors
  to an SD card using the SD library.

  The circuit:
   analog sensors on analog ins 0, 1, and 2
   SD card attached to SPI bus as follows:
 ** MOSI - pin 11
 ** MISO - pin 12
 ** CLK - pin 13
 ** CS - pin 4 (for MKRZero SD: SDCARD_SS_PIN)

  created  24 Nov 2010
  modified 9 Apr 2012
  by Tom Igoe

  This example code is in the public domain.

*/

#include <SPI.h>
#include <SD.h>

const int chipSelect = PIN_Pxn;

void setup() {
  // Open Serial3 communications and wait for port to open:
  Serial3.begin(115200);
  while (!Serial3) {
    ; // wait for Serial3 port to connect. Needed for native USB port only
  }


  Serial3.print("Initializing SD card...");

  // see if the card is present and can be initialized:
  if (!SD.begin(chipSelect)) {
    Serial3.println("Card failed, or not present");
    // don't do anything more:
    while (1);
  }
  Serial3.println("card initialized.");
}

void loop() {
  // make a string for assembling the data to log:
  String dataString = "";

  // read three sensors and append to the string:
  for (int analogPin = 0; analogPin < 3; analogPin++) {
    int sensor = analogRead(analogPin);
    dataString += String(sensor);
    if (analogPin < 2) {
      dataString += ",";
    }
  }

  // open the file. note that only one file can be open at a time,
  // so you have to close this one before opening another.
  File dataFile = SD.open("datalog.txt", FILE_WRITE);

  // if the file is available, write to it:
  if (dataFile) {
    dataFile.println(dataString);
    dataFile.close();
    // print to the Serial3 port too:
    Serial3.println(dataString);
  }
  // if the file isn't open, pop up an error:
  else {
    Serial3.println("error opening datalog.txt");
  }
}`

And I get the following error: 'PIN_Pxn' was not declared in this scope
which is not there when I try it with any other pin declaration from the documentation, but then it does not find the SD Card

**Reply by M65649** [reply]
Are you using this one: https://learn.adafruit.com/adafruit-adalogger-featherwing/pinouts? If you scroll down on the page you see which physical pin the chip select is. On the AVR-IoT Cellular Mini, one can see that pin corresponds to `PB5` on the right (in code that would be `PIN_PB5`):

![Capture](https://github.com/microchip-pic-avr-solutions/avr-iot-cellular-arduino-library/assets/109146295/3105fef0-c05b-49bf-937b-40a02ddcad3d)


We have an example on how to configure pins with the slightly different pin system [here](https://github.com/microchip-pic-avr-solutions/avr-iot-cellular-arduino-library/blob/main/examples/gpio/gpio.ino).

**Reply by vonrle** [reply]
Hi, 
thank you for the fast answer! Yes I'm using that product from Featherwing. And I tried that PB5 Pin but it is not working, sadly. Could be a bad hardware connection, but for the chip select Pin I found in the documentation 
![image](https://github.com/microchip-pic-avr-solutions/avr-iot-cellular-arduino-library/assets/146349576/0746d0ed-dc23-4d4a-bfaf-3ed2a23e0585)
Why is it explicitly stated there? 
But I'll have a look at my hardware and try again with PB5.

**Reply by M65649** [reply]
I think it's explicitly stated there that one can in practice use any pin as chip select. What error are you getting?

**Reply by vonrle** [reply]
Okay, then I didn't read all the guide correctly. I'm not getting any error, but it just can't find the Card. "Card failed, or not present" is given out from my code. because it can not begin the SD. But as I said, maybe then I should have a look at my hardware connection. Or do I have to somehow give the library other Pins for MOSI MISO and SCK, as they are called differently for the AVR Board.

**Reply by M65649** [reply]
Are you using the included SD library from DxCore or another one?

**Reply by vonrle** [reply]
the one included in the DxCore

**Reply by M65649** [reply]
Could you try to run the `CardInfo.ino` sketch to see if it pops out any more information: https://github.com/SpenceKonde/DxCore/blob/master/megaavr/libraries/SD/examples/CardInfo/CardInfo.ino?

**Reply by vonrle** [reply]
I tried, but sadly it also does not find the SD Card and gives out: initialization failed. Things to check:
* is a card inserted?
* is your wiring correct?
* did you change the chipSelect pin to match your shield or module?

**Reply by M65649** [reply]
I did a quick test with the `CardInfo.ino` file and got the following output at least, so the pin setup should be correct. I don't have that SD feather wing, so I can't reproduce your setup. Do you have an oscilloscope or a logic analyzer to check if you get anything on the `MISO` line?

![Capture](https://github.com/microchip-pic-avr-solutions/avr-iot-cellular-arduino-library/assets/109146295/c91f2ff5-b219-4372-b0af-65e0629acae8)

**Reply by vonrle** [reply]
I don't have one, but I can borrow one tomorrow or the day after. I'll let you know, when I got the results. 
But thank you so much for your help!

**Reply by M65649** [reply]
Hello @vonrle,

Did you get the board to recognize the SD card?

**Reply by vonrle** [reply]
Hello M65649, 
thanks for asking. I totally forgot to reply again. The SD Card works now, but I can not figure out how to use the RTC that is on the same shield. It also doesn't find it. Do you have any suggestions what I could try to make it work? It's this shield I'm using: https://learn.adafruit.com/adafruit-adalogger-featherwing/using-the-real-time-clock

Thank you so much for your help!

**Reply by M65649** [reply]
Have you followed this guide: https://learn.adafruit.com/adafruit-adalogger-featherwing/adafruit2-rtc-with-arduino ? The feather is on the I2C 1 bus (see page 7 here: https://ww1.microchip.com/downloads/en/DeviceDoc/AVR-IoT-Cellular-Mini-HW-UserGuide-DS50003320.pdf), so you need to use `Wire1` instead of `Wire`. I had a quick look at the RTCLib, and I believe the begin should be changed to this:

```c
rtc.begin(&Wire1);
```

**Reply by vonrle** [reply]
Yes, I followed the guide and it works with a Feather M0 Adalogger, but not with the AVR. I'm not sure, weather it's because of the diffenrent names for the pins for the AVR and the Feather, that the RTClib can not deal with.

**Reply by M65649** [reply]
See my updated comment above. I believe it is just because we have two I2C buses on the AVR-IoT Cellular Mini.

**Reply by vonrle** [reply]
Perfect, it works now. Thank you so much!

**Reply by M65649** [reply]
Superb 😄
